substrate/staking/pallet/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+    ) -> DispatchResult {
+      let account = ensure_signed(origin)?;
+
+      // add to amount bonded
+      Self::allocate_internal(&account, amount)?;
+
+      // add to participants list for the network
+      let result = VsPallet::<T>::add_participant(account, Amount(amount), network);
+      if result.is_err() {
+        Self::deallocate_internal(&account, amount).unwrap();
+        return result;
+      }
+
+      Ok(())
+    }
+
+    /// Allocate `amount` to a given validator set.
+    #[pallet::call_index(3)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn deallocate(
+      origin: OriginFor<T>,
+      network: NetworkId,
+      #[pallet::compact] amount: u64,
+    ) -> DispatchResult {
+      let account = ensure_signed(origin)?;
+
+      // remove the participant if necessary.
+      // we can't directly deallocate here, since the leaving validator
+      // will be removed after the next session. We only deallocate then
+      // on `end_session` for the right index.
+      VsPallet::<T>::maybe_remove_participant(account, Amount(amount), network)
+    }
+  }
+
+  /// Call order is end_session(i - 1) -> start_session(i) -> new_session(i + 1)
+  /// new_session(i + 1) is called immediately after start_session(i) returns then
+  /// we wait until the session ends then get a call to end_session(i) and so on.
+  impl<T: Config> SessionManager<T::ValidatorId> for Pallet<T> {
+    fn new_session(new_index: u32) -> Option<Vec<T::ValidatorId>> {
+      let next_validators = VsPallet::<T>::next_validator_set(new_index, NetworkId::Serai);
+
+      // Returning None will keep the previous set going.
+      if next_validators.is_empty() {
+        return None;
+      }
+
+      Some(next_validators)
+    }
+
+    fn new_session_genesis(_: u32) -> Option<Vec<T::ValidatorId>> {
+      // this function will be called for index 0 & 1, not just 0.
+      // we return the same set to effectively say that we want
+      // the same validators for sessions 0 & 1.
+      Some(VsPallet::<T>::genesis_validator_set(NetworkId::Serai))
+    }
+
+    fn end_session(end_index: u32) {
+      // do the deallocation of those validator funds
+      // who will not be in the set next session.
+      let key = ValidatorSet { session: Session(end_index + 1), network: NetworkId::Serai };
+      let deallocating_validators = VsPallet::<T>::deallocating_validators(key);
+      for (account, amount, _) in deallocating_validators {
+        // we can unwrap because we are not deallocating more than allocated.
+        <Self as AllocatedStaking<T>>::deallocate(&account, amount.0).unwrap();
+      }
+
+      VsPallet::<T>::end_session(end_index, NetworkId::Serai);
+    }
+
+    fn start_session(start_index: u32) {
+      let validators = SessionPallet::<T>::validators();
+      VsPallet::<T>::start_session(start_index, NetworkId::Serai, validators)
+    }
+  }
+
+  impl<T: Config> AllocatedStaking<T> for Pallet<T> {
```

Comment:
I was actually wondering that myself lol. Why do we have it? it was just there in the original staking folder so I just imported it lol. But will delete because doesn't seem necessary now 👍

Reply:
uhh but that makes the whole `staking/primitives` crate empty? Should we nuke the whole crate then?